tests/installation/ipxe_install.pm

```diff
@@ -141,50 +173,54 @@ END_BOOTSCRIPT
     diag "$response->{status} $response->{reason}\n";
 }
 
+sub enter_o3_ipxe_boot_entry {
+    ipmitool('chassis power reset') unless check_screen([qw(o3-ipxe-menu ipxe-boot-failure)], 180);
+    assert_screen('o3-ipxe-menu', 210);
+    my $key = get_var('HOST_INSTALL_AUTOYAST') ? (is_kvm_host ? 'k' : 'z') : 't';
+    send_key "$key";
+    # try one more time as sometimes sending key does not take effect
+    send_key "$key" if check_screen('o3-ipxe-menu');
+    # confirm the dialog asking for user modifications to kernel, cmdline and initrd
+    # set limit to 10 in case some keys don't go through
+    for (1 .. 10) {
+        last if check_screen([qw(load-linux-kernel load-initrd)], 3);
+        send_key "ret";
```

Comment:
Is this logic correct? This will keep sending the "Enter" keypress blindly even if no dialog shows up.

Reply:
@mdoucha do we need a timeout?

my $max_attempts = 10;
my $timeout = 3;

for (1 .. $max_attempts) {
    if (check_screen([qw(load-linux-kernel load-initrd)], $timeout)) {
        last;
    }
    send_key "ret";
}

Reply:
3 secs is already there

Reply:
No. I think the logic should look like this:

    send_key "ret" if check_screen("needle-checking-whether-the-dialog-showed-up", 30);

The jobs using the ipxe_install module usually run with 3x timeout multiplier so the key will be sent over up to 90 seconds. If the dialog doesn't show up and interactive YaST starts quickly, this'll end up advancing the installation several steps and breaking subsequent installation modules.

Reply:
this logic was introduced by https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17571. It is to adapt the new iPXE menu. the new iPXE menu need user's input to to start the command. "ret" will be sent until it is confirmed that the bootscript has been executed.

see this screenshot: https://openqa.opensuse.org/tests/3537864#step/ipxe_install/6

"ret" keys are required in the 'edit' mode, actually finally 3 "ret" keys are typed in this test.

```
[2023-08-31T11:28:52] no match: -0.7s, best candidate: ipxe_install-o3-ipxe-menu-20220811 (0.00)
[2023-08-31T11:28:52] >>> testapi::_check_backend_response: match=o3-ipxe-menu timed out after 0 (check_screen)
[2023-08-31T11:28:52] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:186 called testapi::check_screen
[2023-08-31T11:28:52] <<< testapi::check_screen(mustmatch=[
    "load-linux-kernel",
    "load-initrd"
  ], timeout=3)
[2023-08-31T11:28:53] no match: 8.2s, best candidate: ipxe_install-load-initrd-20220629 (0.00)
...
[2023-08-31T11:29:02] no match: -0.8s, best candidate: ipxe_install-load-linux-kernel-20220629 (0.79)
[2023-08-31T11:29:021] >>> testapi::_check_backend_response: match=load-initrd,load-linux-kernel timed out after 9 (check_screen)
[2023-08-31T11:29:02] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:187 called testapi::send_key
[2023-08-31T11:29:02] <<< testapi::send_key(key="ret", wait_screen_change=0)
[2023-08-31T11:29:03] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:186 called testapi::check_screen
[2023-08-31T11:29:03] <<< testapi::check_screen(mustmatch=[
    "load-linux-kernel",
    "load-initrd"
  ], timeout=3)
[2023-08-31T11:29:03] no match: 8.7s, best candidate: ipxe_install-load-initrd-20220629 (0.00)
...
[2023-08-31T11:29:12] no match: -0.3s, best candidate: ipxe_install-load-linux-kernel-20220629 (0.79)
[2023-08-31T11:29:12] >>> testapi::_check_backend_response: match=load-initrd,load-linux-kernel timed out after 9 (check_screen)
[2023-08-31T11:29:12] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:187 called testapi::send_key
[2023-08-31T11:29:12] <<< testapi::send_key(key="ret", wait_screen_change=0)
[2023-08-31T11:29:13] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:186 called testapi::check_screen
[2023-08-31T11:29:131] <<< testapi::check_screen(mustmatch=[
    "load-linux-kernel",
    "load-initrd"
  ], timeout=3)
...
[2023-08-31T11:29:22] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:187 called testapi::send_key
[2023-08-31T11:29:22] <<< testapi::send_key(key="ret", wait_screen_change=0)
[2023-08-31T11:29:23] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:186 called testapi::check_screen
[2023-08-31T11:29:23] <<< testapi::check_screen(mustmatch=[
    "load-linux-kernel",
    "load-initrd"
  ], timeout=3)
[2023-08-31T11:29:23] no match: 8.7s, best candidate: ipxe_install-load-initrd-20220629 (0.00)
[2023-08-31T11:29:25] no match: 6.7s, best candidate: ipxe_install-load-initrd-20220629 (0.00)
[2023-08-31T11:29:26] >>> testapi::_handle_found_needle: found ipxe_install-load-initrd-20220629, similarity 1.00 @ 87/386
```

Reply:
the logic works now. Let's fix it in another PR when we find it bring troubles.

Reply:
It seems to me @mdoucha comment is more about accurately handle that operating of sending ret and make test more efficient than blindly trying, which makes sense IMO. @Julie-CAO Do you think this can be improved in future PR by using some needle like needle-checking-whether-the-dialog-showed-up? To me, this is a good thing but minor and not directly related with the main scope of the PR, better to merge it first and improve that later.

If no against comment today, will merge first.

Reply:
Yes, this part was about the new iPXE menu just introduced days ago. It is out of the scope of my PR. And it involves the tests in O3 only. I don't have plan to improve it in this PR.

About if it will be improved in the future, we need monitor more tests before give a solution, because 3 more needles("needle-checking-whether-the-dialog-showed-up")  from SOL console are required. If it is more efficient actually, we need more considerations.

Reply:
OK, future PR is good enough.